### PR TITLE
feat: Implement pre_upgrade and post_upgrade functionality

### PR DESCRIPTION
**Pull request by oblique**

### CODE REVIEW

File: src/ethereum-canister/src/stable_memory.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+/// If memory of `id` is already initialized then that content will be used, otherwise it is
+/// initialized with `Default::default()` value.
+pub(crate) fn init_stable_cell_default<T>(id: MemoryId) -> StableCell<T>
+where
+    T: Storable + Default,
+{
+    init_stable_cell(id, T::default())
+}
+
+pub(crate) fn save_static_string(
```
Comment (zvolin) [inline src/ethereum-canister/src/stable_memory.rs:41]: aren't `static`s there a typo? shouldn't they be `stable`?

Comment (oblique) [reply src/ethereum-canister/src/stable_memory.rs:41]: Not a typo. I wanted this helper to take `std::thread::LocalKey`, which is used when you define a static variable in `thread_local`

Comment (oblique) [reply src/ethereum-canister/src/stable_memory.rs:41]: I can change it to `save_stable_string` but I want to keep `LocalKey`

Comment (zvolin) [reply src/ethereum-canister/src/stable_memory.rs:41]: icool, was just curious

**Review by zvolin — APPROVED** [review]
nice

**Pull request merged**